### PR TITLE
jsonutil/safejson: introducing safejson.String & safejson.Paragraph

### PR DESCRIPTION
**Pull request by chipaca**
In some cases snapd is given a string it can't be sure is safe to print to
the user's terminal.

With this change, jsonutil will have the tools to deal with that
safely.  store/details.go and store/details_v2.go use it to decode
everything the store sends.

### CODE REVIEW

File: jsonutil/puritan/puritan.go [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+	s, err := unmarshal(in, uOpt{basic: true})
+	if err != nil {
+		return err
+	}
+	un.s = s
+	return nil
+}
+
+// Clean returns the string.
+func (un SimpleString) Clean() string {
```
Comment (bboozzoo) [inline jsonutil/puritan/puritan.go]:
```
// Clean returns the string with all non-ASCII characters removed
```

File: jsonutil/puritan/puritan.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+			j := i + sz
+			if r > 0x9f && r != unicode.ReplacementChar && !unicode.Is(strutil.Ctrl, r) {
+				out = append(out, in[i:j]...)
+			}
+			i = j
+		}
+	}
+
+	out = out[:len(out):len(out)]
+	return *(*string)(unsafe.Pointer(&out)), nil
```
Comment (bboozzoo) [inline jsonutil/puritan/puritan.go]: It'd be nice to add a comment here.

File: jsonutil/puritan/puritan.go (hunk 1 above)
Comment (chipaca) [reply jsonutil/puritan/puritan.go]: as discussed in-person :-D, saying this would be misleading; I've added a comment to UnmarshalJSON to reiterate things.

File: jsonutil/puritan/puritan.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+	for i := range un {
+		out[i] = un[i].Clean()
+	}
+
+	return out
+}
+
+// OldPriceMap is a map for prices in the old API.
+type OldPriceMap struct {
+	s map[string]float64
```
Comment (pedronis) [inline jsonutil/puritan/puritan.go]: s/s/m/ ?

File: jsonutil/puritan/puritan.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+// Clean returns a Cleaned map[string]float64.
+func (un OldPriceMap) Clean() map[string]float64 {
+	return un.s
+}
+
+// PriceMap is a map used for prices.
+//
+// the keys are ISO currency codes; the values are strings of floats.
+type PriceMap struct {
+	s map[string]string
```
Comment (pedronis) [inline jsonutil/puritan/puritan.go]: s/s/m/ ?

File: jsonutil/puritan/puritan.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+	for _, r := range s {
+		if r < 'A' || r > 'Z' {
+			return false
+		}
+	}
+	return true
+}
+
+func okPrice(s string) bool {
+	// limits to what strconv.ParseFloat accepts
```
Comment (pedronis) [inline jsonutil/puritan/puritan.go]: could probably be stricter but seems ok for now

File: jsonutil/puritan/puritan.go [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+				// the spec says just ", / and \ can be backslash-escaped
+				// but go adds ' to the list (why?!)
+				out = append(out, in[i])
+				i++
+			default:
+				return "", fmt.Errorf(`unknown escape '%c' at %d of "%s"`, in[i], i, in)
+			}
+		case c <= 0x7e:
+			// printable ASCII, except " or \
+			out = append(out, c)
```
Comment (pedronis) [inline jsonutil/puritan/puritan.go]: is something confusing the pretty printer of GH?  I see the append line in gray

File: jsonutil/puritan/puritan.go [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+		case c < 0x20:
+			return "", fmt.Errorf("unexpected control character at %d in %q", i, in)
+		case c == '\\':
+			// handle escapes
+			i++
+			if i == len(in) {
+				return "", fmt.Errorf("unexpected end of string (trailing backslash) in \"%s\"", in)
+			}
+			switch in[i] {
+			case 'u':
```
Comment (pedronis) [inline jsonutil/puritan/puritan.go]: it this case a copy? or does it need careful reviewing?  it's glazing over prone at first pass

File: jsonutil/puritan/puritan.go [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+			}
+		case c <= 0x7e:
+			// printable ASCII, except " or \
+			out = append(out, c)
+			i++
+		default:
+			r, sz = utf8.DecodeRune(in[i:])
+			j := i + sz
+			if r > 0x9f && r != unicode.ReplacementChar && !unicode.Is(strutil.Ctrl, r) {
+				out = append(out, in[i:j]...)
```
Comment (pedronis) [inline jsonutil/puritan/puritan.go]: why above it seems we error on ctrl chars but here we ignore them (or so it seems) ?

File: store/details.go [hunk 9]
```diff
@@ -23,95 +23,96 @@ import (
...
 type channelSnapInfoDetails struct {
-	Revision     int        `json:"revision"` // store revisions are ints starting at 1
-	Confinement  string     `json:"confinement"`
-	Version      string     `json:"version"`
-	Channel      string     `json:"channel"`
-	Epoch        snap.Epoch `json:"epoch"`
-	DownloadSize int64      `json:"binary_filesize"`
-	Info         string     `json:"info"`
+	Revision     int                  `json:"revision"` // store revisions are ints starting at 1
+	Confinement  snap.ConfinementType `json:"confinement"`
```
Comment (pedronis) [inline store/details.go]: have we rechecked why it was a string?

File: store/details.go [hunk 10]
```diff
@@ -23,95 +23,96 @@ import (
...
 	// TODO: have the store return a 'developer_username' for this
-	Developer   string `json:"origin"`
-	DeveloperID string `json:"developer_id"`
+	Developer   puritan.SimpleString `json:"origin"`
+	DeveloperID puritan.SimpleString `json:"developer_id"`
 
-	Private     bool   `json:"private"`
-	Confinement string `json:"confinement"`
+	Private     bool                 `json:"private"`
+	Confinement snap.ConfinementType `json:"confinement"`
```
Comment (pedronis) [inline store/details.go]: have we rechecked why it was a string?

File: jsonutil/puritan/puritan.go [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+	u, err := strconv.ParseUint(string(in[2:6]), 16, 32)
+	if err != nil {
+		return -1
+	}
+	return rune(u)
+}
+
+type uOpt struct {
+	nlok  bool
+	basic bool
```
Comment (pedronis) [inline jsonutil/puritan/puritan.go]: why basic ? if it's used by simple

File: jsonutil/puritan/puritan.go (hunk 6 above)
Comment (niemeyer) [reply jsonutil/puritan/puritan.go]: Yes, likely the backslash at the end is being seen as escaping the end of line.

**Comment by pedronis** [comment]
as we discussed I'm not quite sure how we scale this to all input from store and snap.yaml if we really need that and also how to avoid string fields getting used by mistake/convenience

File: jsonutil/puritan/puritan.go [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func u4(in []byte) rune {
+	if len(in) < 6 || in[0] != '\\' || in[1] != 'u' {
+		return -1
```
Comment (niemeyer) [inline jsonutil/puritan/puritan.go]: In the name of all the times our ancestors got this wrong in C :slightly_smiling_face: , can we please make this `return 0, false`?

File: jsonutil/puritan/puritan.go [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+			break
+		}
+		i++
+	}
+	if i == len(in) {
+		// wee
+		return string(in), nil
+	}
+	if o.basic {
+		return "", fmt.Errorf("invalid simple JSON string %q: not simple at %d", in, i)
```
Comment (niemeyer) [inline jsonutil/puritan/puritan.go]: I'm afraid this will look a bit strange to people. I mean, it's not like " is not simple.. :)

Instead of describing the terms we used in the implementation, we can just say something like:
```
("character %q in string %q is unsupported for this value", in[i], in)
```

File: jsonutil/puritan/puritan.go [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+	}
+	// in[i] is the first problematic one
+	out := make([]byte, i, len(in)+2*utf8.UTFMax)
+	copy(out, in)
+	var r, r2 rune
+	var sz int
+	var c byte
+	var ubuf [utf8.UTFMax]byte
+	for i < len(in) {
+		switch c = in[i]; {
```
Comment (niemeyer) [inline jsonutil/puritan/puritan.go]: s/;//

File: jsonutil/puritan/puritan.go [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+			if r > 0x9f && r != unicode.ReplacementChar && !unicode.Is(strutil.Ctrl, r) {
+				out = append(out, in[i:j]...)
+			}
+			i = j
+		}
+	}
+
+	out = out[:len(out):len(out)]
+	// This avoids copying the byte array:
+	return *(*string)(unsafe.Pointer(&out)), nil
```
Comment (niemeyer) [inline jsonutil/puritan/puritan.go]: It feels strange that we're resorting to unsafe logic when the entire community is using the json package as is for use cases which are much more intensive than our trivial load. We ought to be able to avoid it as well, at least for this.

File: jsonutil/puritan/puritan.go [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+				// do nothing
+				i++
+			case 'n':
+				if o.nlok {
+					out = append(out, '\n')
+				}
+				i++
+			case '"', '/', '\\':
+				// the spec says just ", / and \ can be backslash-escaped
+				// but go adds ' to the list (why?!)
```
Comment (niemeyer) [inline jsonutil/puritan/puritan.go]: At least here it doesn't. Trying to escape ' results in an error on Unmarshal:
```
invalid character '\'' in string escape code
```

File: jsonutil/puritan/puritan.go [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+			default:
+				return "", fmt.Errorf(`unknown escape '%c' at %d of "%s"`, in[i], i, in)
+			}
+		case c <= 0x7e:
+			// printable ASCII, except " or \
+			out = append(out, c)
+			i++
+		default:
+			r, sz = utf8.DecodeRune(in[i:])
+			j := i + sz
```
Comment (niemeyer) [inline jsonutil/puritan/puritan.go]: That might be nicer as "size".. or just "n".. sz tastes like something else, even more in the context of unicode which has pairs of chars for many unclear purposes. There's at least one more entry above as well.

File: jsonutil/puritan/puritan.go [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+			if r > 0x9f && r != unicode.ReplacementChar && !unicode.Is(strutil.Ctrl, r) {
+				out = append(out, in[i:j]...)
+			}
+			i = j
+		}
+	}
+
+	out = out[:len(out):len(out)]
```
Comment (niemeyer) [inline jsonutil/puritan/puritan.go]: Why is that necessary?

File: jsonutil/puritan/puritan_test.go [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+type escapeSuite struct{}
+
+var _ = check.Suite(escapeSuite{})
+
+var table = map[string]string{
+	"null":           "",
+	`"hello"`:        "hello",
+	`"árbol"`:        "árbol",
+	`"\u0020"`:       " ",
+	`"\uD83D\uDE00"`: "😀",
```
Comment (niemeyer) [inline jsonutil/puritan/puritan_test.go]: Is that right? Do we want emojis in these fields?

File: jsonutil/puritan/puritan.go (hunk 16 above)
Comment (chipaca) [reply jsonutil/puritan/puritan.go]: right, it's a reference to `unquoteBytes` (called out above), which does: https://golang.org/src/encoding/json/decode.go#L1238
there's obviously more going on over there :-)

File: jsonutil/puritan/puritan.go (hunk 18 above)
Comment (chipaca) [reply jsonutil/puritan/puritan.go]: because `cap(out)` could be quite a bit bigger than `len(out)` and as it then does the pointer magic thing, it'd be dragging that extra capacity around

File: jsonutil/puritan/puritan.go (hunk 15 above)
Comment (chipaca) [reply jsonutil/puritan/puritan.go]: Yeah, you're probably right.
OTOH I learend of this way of doing things via https://golang.org/pkg/strings/#Builder.String

File: jsonutil/puritan/puritan_test.go (hunk 19 above)
Comment (chipaca) [reply jsonutil/puritan/puritan_test.go]: do you want to tell our devs they _can't_ use emoji in descriptions?

File: store/details.go (hunk 10 above)
Comment (chipaca) [reply store/details.go]: whatever it was, it's no longer needed (everything works fine without). Likely for transition.

File: jsonutil/puritan/puritan_test.go (hunk 19 above)
Comment (niemeyer) [reply jsonutil/puritan/puritan_test.go]: I want for it to at least be a conscious decision and not a side-effect of not talking about it.

File: jsonutil/puritan/puritan.go (hunk 14 above)
Comment (chipaca) [reply jsonutil/puritan/puritan.go]: that changes the meaning of it though. I'll move it out of the switch.

File: jsonutil/puritan/puritan_test.go (hunk 19 above)
Comment (chipaca) [reply jsonutil/puritan/puritan_test.go]: for the record, I think we _do_ want to allow emojis in descriptions. Disallowing it feels overly prescriptive of us.

There are bound to be things using emojis as part of their name, for example.

File: strutil/ctrl16.go [hunk 20]
```diff
@@ -0,0 +1,51 @@
+// +build !go1.7
```
Comment (zyga) [inline strutil/ctrl16.go]: Sorry for a perhaps naive question, why do we need two versions of this file? Can you please add that as a comment.

File: store/details_v2.go [hunk 21]
```diff
@@ -89,9 +90,9 @@ func infoFromStoreSnap(d *storeSnap) (*snap.Info, error) {
 	info.RealName = d.Name
 	info.Revision = snap.R(d.Revision)
 	info.SnapID = d.SnapID
-	info.EditedTitle = d.Title
-	info.EditedSummary = d.Summary
-	info.EditedDescription = d.Description
+	info.EditedTitle = d.Title.Clean()
```
Comment (zyga) [inline store/details_v2.go]: Could we use String() instead of clean to make this more hassle-free?

File: jsonutil/safejson/safejson.go [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+	s string
+}
+
+func (str *String) UnmarshalJSON(in []byte) (err error) {
+	str.s, err = unmarshal(in, uOpt{})
+	return
+}
+
+// Clean returns the string, with any control characters stripped.
+func (str String) Clean() string {
```
Comment (zyga) [inline jsonutil/safejson/safejson.go]: I suggested to make this String() below, let me know if this is insane in good or bad way please.

File: jsonutil/safejson/safejson.go [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"unicode"
+	"unicode/utf16"
+	"unicode/utf8"
+
+	"github.com/snapcore/snapd/strutil"
+)
```
Comment (zyga) [inline jsonutil/safejson/safejson.go]: Can you please add a comment block explaining why this is needed, how String, Paragraph and other things interact please?

File: jsonutil/safejson/safejson.go [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+	}
+	return rune(u), true
+}
+
+type uOpt struct {
+	nlOK   bool
+	simple bool
+}
+
+func unmarshal(in []byte, o uOpt) (string, error) {
```
Comment (zyga) [inline jsonutil/safejson/safejson.go]: I specifically didn't review this yet. I will return here.

File: store/details_v2.go (hunk 21 above)
Comment (chipaca) [reply store/details_v2.go]: I'm not sure how it would make it more hassle-free; it'd make it more susceptible to using it as a Stringer and passed to, say, `fmt.Sprint()`, but as `Clean()` could in theory be expensive, I thought it best to explicitly avoid that.

Promising that the cleaning will always be done upfront would be another approach, but I'm not sure it's better.

Comment (zyga) [reply store/details_v2.go]: We could make String() call Clean() and cache it, this is a read-only thing after all.

Comment (chipaca) [reply store/details_v2.go]: We could. What's the advantage of that?

File: jsonutil/safejson/safejson.go [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+	"fmt"
+	"strconv"
+	"unicode"
+	"unicode/utf16"
+	"unicode/utf8"
+
+	"github.com/snapcore/snapd/strutil"
+)
+
+// String accepts any valid JSON string
```
Comment (pedronis) [inline jsonutil/safejson/safejson.go]: the docs are strange, because Paragraph/String have the same doc,  they should really explain the intention/usage for them

**Comment by pedronis** [comment]
thanks, docs look good
